File: concurrency-spinlock/src/main/java/io/github/hyeonic/concurrencyspinlock/application/ProductService.java
package io.github.hyeonic.concurrencyspinlock.application;

import io.github.hyeonic.concurrencyspinlock.domain.Product;
import io.github.hyeonic.concurrencyspinlock.domain.ProductRepository;
import java.util.NoSuchElementException;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
public class ProductService {

    private final ProductRepository productRepository;

    public ProductService(final ProductRepository productRepository) {
        this.productRepository = productRepository;
    }

    @Transactional
    public void purchase(final Long id, final Long quantity) {
        var foundProduct = getProduct(id);
        foundProduct.decrease(quantity);
    }

    private Product getProduct(final Long id) {
        return productRepository.findById(id)
                .orElseThrow(NoSuchElementException::new);
    }
}

File: redis-pub-sub/src/main/kotlin/c/hyeoni/redispubsub/domain/product/ProductService.kt
package c.hyeoni.redispubsub.domain.product

import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class ProductService(
    private val productRepository: ProductRepository
) {

    @Transactional
    fun save(product: Product): Product {
        return productRepository.save(product)
    }

    @Transactional
    fun update(id: Long, name: String, description: String): Product {
        return getById(id).apply {
            this.name = name
            this.description = description
        }.also {
            it.notifyChanged()
            productRepository.save(it)
        }
    }

    @Transactional(readOnly = true)
    fun getById(id: Long): Product {
        return productRepository.findById(id)
            .orElseThrow()
    }

    @Transactional(readOnly = true)
    fun findAll(): List<Product> {
        return productRepository.findAll()
    }
}
